### PR TITLE
Added option to validate flags (take 2)

### PR DESCRIPTION
**Pull request by aragilar**
This is an updated version of #12, with new exceptions raised for the tstop and root conditions (it also allows for being able to easily ignore them also. I also removed some printing, which is not needed if we use exceptions.

### CODE REVIEW

File: scikits/odes/sundials/cvode.pyx [hunk 1]
```diff
@@ -1070,7 +1133,10 @@ cdef class CVODE:
         np_tspan = np.asarray(tspan, dtype=float)
         np_y0    = np.asarray(y0, dtype=float)
 
-        return self._solve(np_tspan, np_y0)
+        soln = self._solve(np_tspan, np_y0)
+        if self._validate_flags:
```
Comment (aragilar) [inline scikits/odes/sundials/cvode.pyx]: Forgot to fix this...

**Comment by aragilar** [comment]
I'll fix this, but I might try to run the new version with some of my code, and resubmit when I feel I've tested it enough.

**Comment by bmcage** [comment]
Ok, Pavol is working on an interupt and reinit function. This is not the same thing as you want to tackle, but there is overlap, see his development branch

**Comment by pplk** [comment]
Hi aragilar, hopping into this conversation a bit late, but anyway... :-)
From what I understand, you want to simplify the returning values of the solver and error values return in an exception, so that it would be easy to see whether the solver finished successfully or not.
Hence from the current (flag, t, y, t_out, y_out) = solve(...) we would have (t, y) = solve(...).
Hopefully I did not  miss something else.
So, I  (unlike Benny:-D) don't incline to the use of exceptions as I fail to see the advantage :-) I don't consider a test 'if flag < 0' (or ' if flag == 0' ) too complicated to see whether the computation was successful or not and also not much more verbose than a try...except block:-) Also, I don't feel like tstop or finding a root should be exceptions as they are not error, but expected state.

About the t_out and y_out values, they can be "hidden" e.g. stored as local variables and accessed via a solver method, but they were meant more as stop values of tstop/root, which I consider to be much more used/needed and so I felt they are worth to be returned as they would probably need to be processed. But I agree, the use case can vary :-D

So, to make t_out and y_out more useful, I added 'interruptfn' option which allows to cummulate the tstop/root points without stopping the solver and so now t_out and y_out also hold array of these points (see the development branch). But, as I said, I don't mind having them hidden, though that would break the compatibility.

**Comment by aragilar** [comment]
So I've been spending the last few days looking at getting the validate flags logic working with Pavol's interruptfn (I think I've got it working, needs more testing though). One thing I did notice was that it's now not possible to get the point (and its values) where the solver fails. Was that intentional?

**Comment by pplk** [comment]
No, it was not intentional, but I am aware of it. It is still a proposal of feature I want to have in, though. But I was waiting for your response about your intentions with 'validate_flags' as adding such features might break it. And also, if we are going to add an alternative API, it should be something we all can agree upon, as in the future I imagine it to become the default. I tried to explain my point of view in the previous post.

**Comment by aragilar** [comment]
My plan with 'validate_flags' branch was to share code I'd had for a
while (I started with a much less clean version in 2012/13 wrapping
solve, which became a separate package which I didn't release, and
which I eventually added to the CVODE class, a cleaned up version
becoming #12), 1) because I found it useful, and others may also, and
2) so I didn't need to keep merging it every time master changed.

I'm not that fussed what happens to it, but it's a result of a number
of annoyances with the current API:
1) The debug info from scikits.odes isn't that useful (printing rather
than stderr or using logging, exceptions also help), or informative
(have to look up what each integer value means)
2) If the code succeeds, I care only about y and t (and t only because
t_in may not equal t), not the flag, or y_err/t_err, I can't simply
pass on the returned values without some processing (how the
validate_flags branch started)
3) Copying validation code between projects is annoying (an
improvement here does not appear there, hence the package)

How about a new API for CVODE.solve, one which gets used when
'validate_flags' or 'interruptfn' (or other similar features) are
used? (I think we're both trying to avoid breaking the API, which
means only so much info can be returned)

The new API for CVODE.solve would return a
[namedtuple](https://docs.python.org/2/library/collections.html#collections.namedtuple),
which can be extended with a number of different values (i.e. you
could have separate values which stored where the error occured, roots
were found and the values at tstop). This also means we can change how
CVODE._solve works (add new values, change order etc.) without
specifying specific values.

**Comment by aragilar** [comment]
See https://github.com/aragilar/odes/tree/new-api for how this could work.

**Comment by pplk** [comment]
I am not emotionally fixated to the old API, so as I said previously I don't mind a new one. But then I expect it to serve us all. And once consensus is reached, it's all systems go :-)

I find your proposal a step to the right direction, it certainly gives us more flexibility than we currently have. But what if someone would use destructuring bind on the tuple (like we use so far) and then we extend it and add some new feature (like my interruptfn) - we break the API for him.
So if to break API, then let's do (preferably) only once.

Unless I am mistaken all flags < 0 are unrecoverable errors. If this would be true for all sundials solvers (who said we can't interface more in the future?:-D), we can simply throw an exception in solve() as you proposed first. The user can still do some of his black magic tricks if he decides to. 
This would then leave us with flags 0 (success), 1 (tstop), 2 (root) and 99 (success with warning).
Not considering flags 1 and 2 (flag 99 is still success) we would have
(t, y) = solver.solve(...)
which is what you want. No problem with this and here we can forget also the 'flag' itself.

But TStop and Root are expected states during/after computation. We use them often and so we need to distinguish between those three states (success, tstop and root) to perform different actions. Moreover (considering my proposal of "interrupfn") I may also want to return all points at which tstop was reached and/or root was found. And about this I still don't have clear picture.
I am thinking about adding
(t_tstop, y_tstop) = solver.tstops()
(t_roots, y_roots) = solver.roots()
which would return all the tstop/root points, only the last one or (None, None) if none encountered.
But would I want to collect all tstop and root points, there would be no way to distinguish whether we stopped at success, root or tstop - well, unless we compare the last time reached of each. Question is: would that be a problem?

@Benny What do you think?

**Comment by bmcage** [comment]
named tuple is nice. The branch does not add much else though. turning flags < 0 in exceptions is nice too. That should help validation code. 
on flag >=0, the named tuple does not change anything. We should find a combined solution for all. .tstops and .roots is something to try to see how it works in practice

**Comment by aragilar** [comment]
The new-api branch was more of an example than anything new. I created a new branch (https://github.com/aragilar/odes/tree/new-api-with-valiate-and-interrupt) where I rebased (in the wrong order, but never mind, it's an idea, not a final version, also, I haven't set some things to None which should be None) the new-api branch onto the validate and interrupt branches.

The idea is we return a namedtuple of the form (flag_name, flag_number, (t, y), (t_err, y_err), (t_roots, y_roots), (t_tstop, y_tstop)), but the order isn't important given we're using a named tuple.

This means if you use validate flags, you could have some code like
```python
solver = ode('cvode', rhs, validate_flags=True)
t, y = solver.solve(t_in, init).values
```
and you'd either get the points you want, or the program exits with an exception telling you what went wrong (or use try/except to deal with the error).

It also makes it easy to get roots/tstop out, and means that if the solver fails later you see where it failed, and what roots were found.

**Pull request closed without merging**